clients/nexus-client/src/lib.rs

```diff
@@ -22,6 +22,11 @@ progenitor::generate_api!(
         slog::debug!(log, "client response"; "result" => ?result);
     }),
     replace = {
+        // It's kind of unfortunate to pull in such a complex and unstable type
+        // as "blueprint" this way, but we have really useful functionality
+        // (e.g., diff'ing) that's implemented on our local type.
+        Blueprint = nexus_types::deployment::Blueprint,
```

Comment:
I share yours and Adam's reluctance about `replace`.  In this case, the functionality being shared (diff'ing blueprints) is pretty non-trivial and I think it would be fairly painful to impl the requisite `From`s: I think it would require implementing `From` for `Blueprint`, and `OmicronZonesConfig`, which in turn requires `OmicronZoneConfig`, then `OmicronZoneType` (a big enum with a bunch of properties in many variants), `OmicronZoneDataset`, `SledRole`, `ZpoolName`, `NetworkInterface`, `NetworkInterfaceKind`, and `SourceNatConfig`.  (I'm not positive, but I think that's right.)

I think better approaches here might be to have Progenitor validate that your `replace` type really is compatible with the spec, or alternatively if there were a way to automatically generate `From` impls between two types that are identical (which would presumably break -- correctly -- if they weren't identical).  It doesn't feel sustainable to me to either duplicate complex functionality on multiple types or hand-roll so many complex `From` impls.  What do you think?

Reply:
I actually had the opposite reaction, and didn't think it was unfortunate to use `replace` at all. However, clearly I am missing what is problematic here.

Reply:
I think what sucks about using `replace` is that you lose the ability to notice when your local type doesn't match the spec.  In turn that means we lose all of the benefit of Rust's strong static typing across the OpenAPI boundary.

Concretely: in the normal case where Progenitor generates the types for you, say we pull in a new version of the spec with a new required field.  Progenitor's type will have that new field in it.  So if we had a code path that creates an instance of that type (i.e., to make the request) it will now fail to compile.  That's good: if we didn't know it before, we now know we've made an incompatible change to the spec.  If we did know it, now's when we update our code to fill in the new value.

If on the other hand we use `replace` to supply the type, when we pull in the updated spec, everything still compiles -- there's no reason it wouldn't.  Instead, that code path will fail at runtime when the server rejects the request because it's missing the required field.  That really sucks.

Similarly on the output side, if we use `replace` to supply our own type, we don't find out at compile time when our type is incompatible.  A particularly bad case would be an enum, where the server might add a new variant (which one might even consider a non-breaking change).  But if the client sees it, it will fail to deserialize the response.  By contrast, if we use the Progenitor-generated type, we'll fail to build if, say, we try to `match` on the enum and don't include the new variant.

In summary, using Progenitor types gets a lot of the benefit of Rust across the OpenAPI boundary.  Using "replace" bypasses all of it (for that type).  The downside of Progenitor types is that when there's non-trivial functionality on the original type, you either have to impl that functionality twice (sucks) or convert the type to your own type (okay for simple types in a few places, but kind of sucks in the large).  For simple, stable types like `Name` I think the benefits of "replace" probably outweigh the risks.  For complex types that are likely to change, I'm less sure (but still feel what I said earlier about this case).

Reply:
> I think better approaches here might be to have Progenitor validate that your replace type really is compatible with the spec

This feels like it would be huge - it would allow use to use `replace` fearlessly, even on big / complex / volatile types like this one, I think?

Reply:
> > I think better approaches here might be to have Progenitor validate that your replace type really is compatible with the spec
> 
> This feels like it would be huge - it would allow use to use `replace` fearlessly, even on big / complex / volatile types like this one, I think?

Potentially, yeah. I spoke with Adam about this a bit.  It's not clear how we could do this at compile time, but I imagine we might be able to do it at test-time.